[SPARK-52772][SQL] Inconsistent table attribute handling during updates

ksbeyer opened the pull request:
### What changes were proposed in this pull request?

TableOutputResolver always adds an Alias with explicit metadata that propagates table attribute metadata to query attributes and prevents query metadata from flowing to the writer.  A couple of rewrites (trimNonTopLevelAliases, removeRedundantAlias) needed to be changed to preserve the correct flow of empty metadata.  This triggered a need to change to Project.doCanonicalize to consider aliases equivalent based on the effective metadata instead of the explicit and non-inheritable metadata.

Passing this table information via the source query attributes is not the best design. Ideally we should change all the writers to pick up table configuration from the table directly. However, there are many third-party connectors that may rely on this behavior.  So for now, we will simply make the query output attributes match the target column's names and metadata in all cases.

### Why are the changes needed?

Fixes bugs where
1. TableOutputResolver was inconsistent in copying the table column metadata to the query attributes. The complex type cases did not propagate the table metadata, but other cases did.
2. Source table metadata, like default value definitions, leaked out to the writer that then incorrectly reacted to the metadata.
3. removeRedundantAlias and trimNonTopLevelAliases incorrectly changed empty metadata to non-empty.  These were only exposed while attempting to fix the above.
4. There was a failing JSON test case in InsertSuite tracked by SPARK-47029: ALTER COLUMN DROP DEFAULT fails to work correctly with JSON data sources.  This test now works. The test used a struct field with a default value in JSON; because of (1) the JSON writer did not find the default value on the query attribute metadata.  If that is the full

### Does this PR introduce _any_ user-facing change?

This ensures that the table column metadata is applied to the query attributes in all cases, and it should prevent source metadata from flowing from the source to the writer.  This will have user-facing changes when writers are sensitive to the query attribute metadata.  This should be fixing bugs and inconsistencies.  For example, when JSON_GENERATOR_WRITE_NULL_IF_WITH_DEFAULT_VALUE is true (which is default), complex type fields with a default written using JSON will now get the requested explicit NULL instead of the field omitted.

### How was this patch tested?

Test cases added to InsertSuite for cases when the metadata did not flow correctly.complex types that failed before changes and passes after.  Also added some test cases to V2WriteAnalysisSuiteBase to validate changes to query attribute metadata did not break things.

### Was this patch authored or co-authored using generative AI tooling?

No

ksbeyer committed:
[SPARK-52772]
TableOutputResolver no longer propagates table attribute metdata to query attributes.
InsertIntoHadoopFsRelationCommand propagates existenceDefaultValue from table
attributes to query attributes only for JSON output.
This fixes cases with complex types not getting explicit nulls in JSON when it should.

ksbeyer commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:
```diff
@@ -180,7 +180,7 @@ object TableOutputResolver extends SQLConfHelper with Logging {
         CharVarcharUtils.stringLengthCheck(casted, attr.dataType)
       }
-      Alias(exprWithStrLenCheck, attr.name)(explicitMetadata = Some(attr.metadata))
+      Alias(exprWithStrLenCheck, attr.name)()
```
The metadata is inconsistently set.  It is not set for complex types, but is set for simple types and for promotions.  Either it needs to be set always or never.

If we transfer all column attributes to the query, then bad things happen with char / varchar metadata -- which is why I think the code was marking non-promotable metadata, which I also removed.

I don't see any solid reason to move column properties onto the query columns.  I'd rather have table columns, and perhaps the entire definition, made available to the writers.  I considered making that change but it requires making changes that affect all v1 writers. I'd like to remove the metadata use altogether, but the change was too big.   Instead, I just pushed the issue into a very special case for JSON and fixed the inconsistent null handling.  I didn't find any more tests failing, so I'm hoping, perhaps naively, that there isn't more (half broken!) need for the metadata.    If there are more uses, they are not working correctly with complex types.  

We can go the other way and make all types return metadata, but it copies all table attribute metadata without a clean understanding of impact. Then we get random attempt to suppress metadata that causes problems (eg varchar supression). It will also make the dsv2 transition harder, as we continue to flow Column configuration via metadata. We need a clear design and documentation on what is transferred and why.

gengliangwang commented:
@ksbeyer can we fix the issue by propagating the `metadata` from `InsertIntoHadoopFsRelationCommand.catalogTable.schema` to `FileFormatWriter.OutputSpec`? This can be a more general fix

ksbeyer commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala:
```diff
@@ -222,6 +227,37 @@ case class InsertIntoHadoopFsRelationCommand(
+   * attribute metadata, but this is a nontrivial change.
+   */
+  private def markColumnsWithDefaultForJson(outputColumns: Seq[Attribute]): Seq[Attribute] = {
+    if (catalogTable.isEmpty || !fileFormat.isInstanceOf[JsonFileFormat] ||
```
My goal is to eliminate the propagation.   I don't think we should propagate the table attribute metadata to query attributes.   I want to limit the propagation for a documented purpose.  Is should also file another jira to fix the json writer to get table info and remove this propagation.

At a broader scale, the use of schema and attribute metadata seems pretty adhoc and error prone.  Eg, I see questionable propagation of char/varchar info that makes it unclear where the length constraints / padding should be enforced.  This forced me to add unnecessary aliases to hide the metdata that otherwise causes the write logic to get upset.  I'm still unsure if / where the constraints are enforced.  I think we should push to eliminate the metadata; perhaps the 1-pass analyzer can help propagate the right information during analysis rather than adhoc tags.

ksbeyer commented:
> @ksbeyer can we fix the issue by propagating the `metadata` from `InsertIntoHadoopFsRelationCommand.catalogTable.schema` to `FileFormatWriter.OutputSpec`? This can be a more general fix

We could add an `tableSchema: Option[StructType]` to the OutputSpec.  To eliminate the option, I could make it be the table schema when the table is provided, else the data schema.

The OutputSpec doesn't flow to the JacksonGenerator, however.  Could/should we change the dataSchema in FileFormatWriter.write to be the table schema (when provided) instead of the query schema?

ksbeyer committed:
[SPARK-52772] Pass table schema (when available) down to file writers in place of the query schema.
This allows the JacksonGenerator to detect coluns with default values without propagating table
schema metadata to query columns.

ksbeyer commented:
New commit is now passing the table schema (when available) down to the file writers, and the json specific hack has been removed.

ksbeyer committed:
[SPARK-52772] Fixed missed OutputSpec constructor call

gengliangwang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:
```diff
@@ -114,17 +116,18 @@ object FileFormatWriter extends Logging {
-    val dataSchema = dataColumns.toStructType
-    DataSourceUtils.verifySchema(fileFormat, dataSchema)
-    DataSourceUtils.checkFieldNames(fileFormat, dataSchema)
+    val tableSchema = finalOutputSpec.tableSchema
```
From line 112:
```
val dataColumns = finalOutputSpec.outputColumns.filterNot(partitionSet.contains)
```
We should filter out the partition columns from `tableSchema` as well

gengliangwang replied:
I mean, the schema here should not include the partition columns. Otherwise it is breaking change and there can be bugs.

ksbeyer committed:
[SPARK-52772] fix scala style error

ksbeyer committed:
[SPARK-52772] Restrict WriteJobDescription.tableSchema to non-partitioning fields.

ksbeyer committed:
[SPARK-52772] Reverted changes to pass table schema to file writers.
Changed TableOutputResolver to consistently apply table attribute names and metadata in all cases.

ksbeyer committed:
[SPARK-52772] Accept minor plan improvement

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:
```diff
@@ -180,12 +180,38 @@ object TableOutputResolver extends SQLConfHelper with Logging {
+   * looks for default value metadata to control some behavior.  We also remove any
+   * remaining char/varchar metadata because it should be handled by this time.
+   * If we don't do this,
+   *     DeltaBasedMergeIntoTableUpdateAsDeleteAndInsertSuite.
```
nit: let's mention the behavior instead of the test suite

cloud-fan replied:
+1

ksbeyer replied:
I rewrote the comments

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:
```diff
@@ -180,12 +180,38 @@ object TableOutputResolver extends SQLConfHelper with Logging {
+   *     org.apache.spark.sql.catalyst.plans.logical.V2WriteCommand.areCompatible
+   * which prevents the write operation from resolving.
+   */
+  private def makeNamed(expr: Expression, attr: Attribute): NamedExpression = {
```
how about rename as `applyColumnMetadata`

ksbeyer replied:
renamed

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:
```diff
@@ -180,12 +180,38 @@ object TableOutputResolver extends SQLConfHelper with Logging {
+   * which prevents the write operation from resolving.
+   */
+  private def makeNamed(expr: Expression, attr: Attribute): NamedExpression = {
+    val requiredMetadata = CharVarcharUtils.cleanMetadata(attr.metadata)
```
nit: let's add a comment about why this is necessary

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:
```diff
@@ -180,12 +180,38 @@ object TableOutputResolver extends SQLConfHelper with Logging {
+    expr match {
+      case expr: NamedExpression if expr.name == attr.name && expr.metadata == requiredMetadata =>
+        expr
+      // Save an Alias if we can change the name directly.
```
This looks risky as we reuse the attribute ID.

ksbeyer replied:
This code has been changed; I got rid of the direct Attribute name/metadata changing and always use a new Alias when I want to change the name or metadata. At least one reason mucking with attributes wasn't a good idea is the RemoveNoopOperators rule was discarding my changes.

cloud-fan commented:
shall we update the PR description? Seems it doesn't match what the PR actually does now.

ksbeyer committed:
[SPARK-52772] Fix handling of empty metadata in removeRedundantAlias and trimNonTopLevelAliases.
Fix applyColumnMetadata to properly flow table column metadata and block source metadata.
Add tests in InsertSuite for cases when source metadata leaked before these changes.

ksbeyer committed:
[SPARK-52772] Fix trimNonTopLevelAliases to preserve the effective metadata explicitly.
Make MetadataBuilder share the empty singleton when build is empty.
Update expected results to remove some unnecessary casts.

ksbeyer committed:
[SPARK-52772] Make Project.doCanonicalize consider an alias equivalent to an attribute if the metadata is equivalent,
instead of considering if the alias has explicit metadata or non-inheritance.

ksbeyer commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:
```diff
@@ -132,7 +132,11 @@ object TableOutputResolver extends SQLConfHelper with Logging {
           tableName, valueType, colType, byName = true, conf, addError, colPath)
-        if (canWriteExpr) checkNullability(value, col, conf, colPath) else value
+        if (canWriteExpr) {
+          applyColumnMetadata(checkNullability(value, col, conf, colPath), col)
```
Correct.  

I'd prefer to remove the table metadata from the query attributes instead, but there was concern about the risk to all the connectors out there that might be using the metadata.  

So this change will make the table metadata _always_ propagate to the query attributes and _only_ the table attributes (removing the source attributes that leaked through sometimes).

ksbeyer commented on sql/core/src/test/resources/sql-tests/analyzer-results/charvarchar.sql.out:
```diff
@@ -376,7 +376,7 @@ CreateDataSourceTableCommand `spark_catalog`.`default`.`char_tbl4`, false
 -- !query analysis
 InsertIntoHadoopFsRelationCommand file:[not included in comparison]/{warehouse_dir}/char_tbl4, false, Parquet, [path=file:[not included in comparison]/{warehouse_dir}/char_tbl4], Append, `spark_catalog`.`default`.`char_tbl4`, org.apache.spark.sql.execution.datasources.InMemoryFileIndex(file:[not included in comparison]/{warehouse_dir}/char_tbl4), [c7, c8, v, s]
-+- Project [static_invoke(CharVarcharCodegenUtils.charTypeWriteSideCheck(cast(c#x as string), 7)) AS c7#x, static_invoke(CharVarcharCodegenUtils.charTypeWriteSideCheck(cast(c#x as string), 8)) AS c8#x, static_invoke(CharVarcharCodegenUtils.varcharTypeWriteSideCheck(cast(v#x as string), 6)) AS v#x, cast(c#x as string) AS s#x]
++- Project [static_invoke(CharVarcharCodegenUtils.charTypeWriteSideCheck(cast(c#x as string), 7)) AS c7#x, static_invoke(CharVarcharCodegenUtils.charTypeWriteSideCheck(cast(c#x as string), 8)) AS c8#x, static_invoke(CharVarcharCodegenUtils.varcharTypeWriteSideCheck(cast(v#x as string), 6)) AS v#x, c#x AS s#x]
```
The cast is avoided by the change in 
   org.apache.spark.sql.catalyst.analysis.TableOutputResolver.checkField
from
   if (isCompatible(tableAttr, queryExpr))
where isCompatible:
    DataTypeUtils.sameType(tableAttr.dataType, queryExpr.dataType) &&
      tableAttr.name == queryExpr.name &&
      tableAttr.metadata == queryExpr.metadata
to
  if (DataTypeUtils.sameType(tableAttr.dataType, queryExpr.dataType))  

The name check is not required because we will make sure the name is correct in applyColumnMetadata.  The metadata check seems wrong. Why do we care if the table side has a default value that matches the query side?  

I'm guessing the concern is solely for the char/varchar metadata.  For this, the cast did help in a complicated way: it helped to keep the char/varchar metadata from propagating to the write command, but it is a no-op cast from String (with char/varchar metadata) to String as far as i can tell.  With the new code, applyColumnMetadata will keep the char/varchar metadata from propagating (along with the rewrite changes to not undo its work).

ksbeyer commented:
> shall we update the PR description? Seems it doesn't match what the PR actually does now.

I updated it, but assuming with proceed with the current PR, the description needs to be updated again...  I decided to fix some rewrites instead of using a hacky metadata tag.  I'll revisit once we're aligned on approach.

ksbeyer committed:
[SPARK-52772] minor cleanup

ksbeyer committed:
[SPARK-52772] re-enable test; changes fix SPARK-47029

cloud-fan commented:
thanks, merged to master!
